Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+The stdError messages are not part of the remoting protocol and need to be handled separately.
+
+### Creating a connection
+
+The PowerShell module implementing the new remote connection should supply all necessary cmdlets for creating and managing the connection.
+In particular it should have a cmdlet that returns a `PSSession` object.
+
+```powershell
+PS C:\> Import-Module -Name MyCustomConnection
+PS C:\> New-MyCustomSession -ComputerName remoteVM1 -Cred $myCredentials
+
+Id Name            Transport ComputerName    ComputerType    State         ConfigurationName     Availability
+ -- ----            --------- ------------    ------------    -----         -----------------     ------------
+  1 Runspace1       MyCustom  remoteVM1       RemoteMachine   Opened        DefaultShell             Available
+```
+
+However, if the new connection info class is made public accessible, the PowerShell remoting API can also be used to create a connection instance.
+
+```powershell
+PS C:\> Import-Module -Name MyCustomConnection
+PS C:\> $connectionInfo = [MyCustomConnection.CustomConnectionInfo]::new($computerName, $creds)
+PS C:\> $runspace = [runspacefactory]::CreateRunspace($connectionInfo)
+PS C:\> $runspace.Open()
+PS C:\> $runspace
+
+ Id Name            ComputerName    Type          State         Availability
+ -- ----            ------------    ----          -----         ------------
+  2 Runspace2       remote-1        Remote        Opened        Available
+```
+
+### Endpoints
+
+A remote connection must include a PowerShell session on the target machine, that handles remoting protocol messages from the client.
+This is called the remoting endpoint.
+An endpoint typically involves a service that listens for client connection requests, and securely creates the PowerShell session under the correct account and with correct permissions.
+For example, both WSMan and SSH have services that do this.
+Endpoints such as these are outside the scope of this feature proposal, and it is assumed that any custom connection scheme will include some sort of endpoint support.  
+
+PowerShell currently supports a 'server' running mode, where remoting protocol messages are channeled through the process stdIn/stdOut data streams.
+Any custom endpoint solution will leverage this to establish a PowerShell session and handle remoting protocol messages.
+Currently, this is the only way to run PowerShell as an endpoint server, and should be sufficient for any custom solution.
+Also this server mode is available on all supported PowerShell versions (v5.1 - v7.1+).
+
+#### Example - PowerShell in server mode
+
+This example illustrates how PowerShell can be run in server mode.
+
+```powershell
+# Create PowerShell endpoint session (with default shell configuration)
+# Protocol handles messages through the child process stdIn/stdOut process pipes
+$proc = Start-Process -FilePath pwsh.exe -ArgumentList '-servermode -nologo -noprofile' -PassThru
```

Comment:
This is a bit of a misnomer, you cannot on Windows (AFAIK) hook into a processes stdio after it is created. You need to specify the pipes at creation to a handle of your choice. `Start-Process` can do this to an extent but only to a file rather than pipes that you control.

This doesn't rule out the named pipe communication I mentioned above though, just the stdio communication mentioned here.